gateway-service-impl/src/main/java/org/hypertrace/gateway/service/baseline/BaselineServiceImpl.java

```diff
@@ -0,0 +1,383 @@
+                BaselineEntity.newBuilder()
+                    .setEntityType(value.getEntityType())
+                    .setId(value.getId())
+                    .putAllBaselineAggregateMetric(value.getBaselineAggregateMetricMap());
+            if (!baselineEntityTimeSeriesMap.isEmpty()) {
+              baselineEntityBuilder.putAllBaselineMetricSeries(
+                  baselineEntityTimeSeriesMap.get(key).getBaselineMetricSeriesMap());
+            }
+            BaselineEntity baselineEntity = baselineEntityBuilder.build();
+            baselineEntityList.add(baselineEntity);
+          });
+    } else if (!baselineEntityTimeSeriesMap.isEmpty()) {
+      baselineEntityTimeSeriesMap.forEach(
+          (key, value) -> {
+            BaselineEntity baselineEntity =
+                BaselineEntity.newBuilder()
+                    .setEntityType(value.getEntityType())
+                    .setId(value.getId())
+                    .putAllBaselineMetricSeries(value.getBaselineMetricSeriesMap())
+                    .build();
+            baselineEntityList.add(baselineEntity);
+          });
+    }
+    return baselineEntitiesResponseBuilder.addAllBaselineEntity(baselineEntityList).build();
+  }
+
+  private Map<String, BaselineEntity> getEntitiesMapFromTimeSeriesResponse(
+      BaselineEntitiesResponse baselineEntitiesResponse, long startTimeMillis) {
+    Map<String, BaselineEntity> baselineEntityMap = new HashMap<>();
+    for (BaselineEntity baselineEntity : baselineEntitiesResponse.getBaselineEntityList()) {
+      Map<String, BaselineMetricSeries> metricSeriesMap =
+          baselineEntity.getBaselineMetricSeriesMap();
+      Map<String, Baseline> baselineMap = new HashMap<>();
+      BaselineEntity.Builder baselineEntityBuilder =
+          BaselineEntity.newBuilder()
+              .setEntityType(baselineEntity.getEntityType())
+              .setId(baselineEntity.getId());
+      Map<String, BaselineMetricSeries> revisedMetricSeriesMap = new HashMap<>();
+      // Calculate baseline
+      metricSeriesMap.forEach(
+          (key, value) -> {
+            List<Double> metricValues =
+                value.getBaselineValueList().stream()
+                    .map(x -> x.getBaseline().getValue().getDouble())
+                    .collect(Collectors.toList());
+            double[] metricValueArray =
+                metricValues.stream().mapToDouble(Double::doubleValue).toArray();
+            Baseline baseline = BaselineCalculator.getBaseline(metricValueArray);
+            baselineMap.put(key, baseline);
+          });
+      // Update intervals
+      metricSeriesMap.forEach(
+          (key, value) -> {
+            List<BaselineInterval> intervalList = value.getBaselineValueList();
+            List<BaselineInterval> revisedList = new ArrayList<>();
+            for (BaselineInterval baselineInterval : intervalList) {
+              if (baselineInterval.getStartTimeMillis() >= startTimeMillis) {
```

Comment:
But aren't baselines based on a time range that extends further into the past (never the future)? So a calculated interval should never be newer than the requested.

Reply:
Suppose user selected last 1 hour we calculate time series for last 24 hours, here we are having this condition to send baseline for last 1 hour data instead of sending all 24 hours.

Reply:
So doesn't that mean this is relying on the incorrect implementation of baseline time range (that is, it overlapping the requested time range)? If I'm reading this right, it's using whatever intervals were requested for the baseline input data (an implementation detail), filtering them, then cloning them as new intervals, stuffing the calculated baseline into each one.

Reply:
Will fix this, as part of updating endTime.